File: biddo_client/lib/core/repositories/favourites.dart
import 'package:get/get.dart';

import '../models/account.dart';
import '../models/auction.dart';
import 'base.dart';

class FavouriteRepository {
  final dio = Get.find<Api>();

  Future<List<Auction>> loadAll() async {
    try {
      var response = await dio.api.get('/favourites');
      return List<Auction>.from(
        response.data.map(
          (el) => Auction.fromJSON(el),
        ),
      );
    } catch (error) {
      print('Error loading favourite auctions: $error');
      return [];
    }
  }

  Future<void> add(String auctionId) async {
    try {
      await dio.api.put('/favourites/add/$auctionId');
    } catch (error) {
      print('error adding auction to favourites: $error');
      return;
    }
  }

  Future<void> remove(String auctionId) async {
    try {
      await dio.api.put('/favourites/remove/$auctionId');
    } catch (error) {
      print('error removing auction from favourites: $error');
      return;
    }
  }

  Future<List<Account>> loadAccountsThatHaveAuctionInFavourites(
    String auctionId, [
    int page = 0,
    int perPage = 20,
  ]) async {
    try {
      var response = await dio.api.get(
        '/favourites/accounts/$auctionId/$page/$perPage',
      );
      return List<Account>.from(
        response.data.map(
          (el) => Account.fromJSON(el),
        ),
      );
    } catch (error) {
      print('error getting accounts that have auction in favourites $error');
      return [];
    }
  }
}

File: biddo_web/src/core/repositories/favourites.ts
import { Account } from '../domain/account'
import { Auction } from '../domain/auction'
import { RequestMaker, RequestType } from '../services/request-maker'

class FavouriteRepository {
  private basePath = '/favourites'

  public async loadAll(): Promise<Auction[]> {
    try {
      const response = (await RequestMaker.makeRequest({
        path: this.basePath,
        method: RequestType.GET,
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Auction.fromJSON(el))
    } catch (error) {
      console.error('Error loading favourite auctions:', error)
      return []
    }
  }

  public async add(auctionId: string): Promise<void> {
    try {
      await RequestMaker.makeRequest({
        path: `${this.basePath}/add/${auctionId}`,
        method: RequestType.PUT,
      })
    } catch (error) {
      console.error('Error adding auction to favourites:', error)
    }
  }

  public async remove(auctionId: string): Promise<void> {
    try {
      await RequestMaker.makeRequest({
        path: `${this.basePath}/remove/${auctionId}`,
        method: RequestType.PUT,
      })
    } catch (error) {
      console.error('Error removing auction from favourites:', error)
    }
  }

  public async loadAccountsThatHaveAuctionInFavourites(
    auctionId: string,
    page = 0,
    perPage = 20
  ): Promise<Account[]> {
    try {
      const response = (await RequestMaker.makeRequest({
        path: `${this.basePath}/accounts/${auctionId}/${page}/${perPage}`,
        method: RequestType.GET,
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Account.fromJSON(el))
    } catch (error) {
      console.error('Error getting accounts that have auction in favourites:', error)
      return []
    }
  }
}

const FavouriteRepositoryInstance = new FavouriteRepository()
export { FavouriteRepositoryInstance as FavouriteRepository }
